### PR TITLE
✨ 51 - add permissions to groups panel

### PR DESCRIPTION
**Pull request by anncatton**
Adds permissions to group panel
- client side sorting and search on `/group/permissions`
- disables sort on `name` field for `/policies/{id}/{users/groups}` until server side fix
- fixes user panel update when api key revoked from child table
- removes item from selector dropdown when selected on edit. hides already selected items
- removes list limit on edit dropdown
- clears staged changes on cancel edit
- corrects columns on tables based on parent and child type
- sets default displayMode to table
- adds clear function to search input
- fixes nav link overriding table rows when nav is collapsed
- disables save when new permissions added without setting accessLevel

### CODE REVIEW

File: src/common/typedefs/Resource.d.ts [hunk 1]
```diff
@@ -123,7 +129,8 @@ export interface IResource {
   noDelete?: true;
   name: { singular: string; plural: TResourceType };
   ListItem: JSX.Element<any>;
-  getList: (params: IListParams) => Promise<IListResponse>;
+  // getList can now be a hashmap
+  getList: any;
```
Comment (hlminh2000) [inline src/common/typedefs/Resource.d.ts]: if it's a hashmap I'd type this as `{[k: string]: WhateverValueTypeHere}`. If it's polymorphic, a union type can be used :)

File: src/common/associatedUtils.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+const isChildOfPolicy = parent => hasParent(parent) && isPolicy(parent);
+const isGroupPermission = (parent, child) =>
+  hasParent(parent) && isGroup(parent) && isPermission(child);
+const isUserPermission = (parent, child) =>
+  hasParent(parent) && isUser(parent) && isPermission(child);
+
+const isPolicy = resource => getResourceName(resource) === 'policy';
+const isPermission = resource => getResourceName(resource) === 'permission';
+const isUser = resource => getResourceName(resource) === 'user';
+const isGroup = resource => getResourceName(resource) === 'group';
```
Comment (hlminh2000) [inline src/common/associatedUtils.ts]: out of curiosity, are `Policy`, `Permission`, `User`, `Group` and `ApiKey` typescript types? if so, these functions can be typed as `(resource: Resource) => resource is Policy`. Probably will break things more than helpful for now so not needed, but just FYI :)

File: src/stateProviders/provideEntity.ts [hunk 3]
```diff
@@ -1,9 +1,16 @@
...
-import { findIndex, omit, uniq } from 'lodash';
+import { findIndex, isEmpty, omit, uniq } from 'lodash';
 
+import { isGroup, isPolicy } from 'common/associatedUtils';
 import RESOURCE_MAP from 'common/RESOURCE_MAP';
 
 const MAX_ASSOCIATED = 5;
+
+export const getListFunc = (associatedType, parent) =>
+  associatedType === 'permissions' && !isEmpty(parent)
```
Comment (hlminh2000) [inline src/stateProviders/provideEntity.ts]: these magic strings `'permissions'` seem to be scattered around quite a lot, should we extract them to an enum or constants somewhere? seems a bit prone to typos

File: src/common/associatedUtils.ts (hunk 2 above)
Comment (anncatton) [reply src/common/associatedUtils.ts]: There are ts types for the entities, but they are more of a DTO...the resource type here is from that giant resource_map file.

File: src/stateProviders/provideEntity.ts (hunk 3 above)
Comment (anncatton) [reply src/stateProviders/provideEntity.ts]: Agree! Will add that

**Pull request merged**